### PR TITLE
Added Tectonic.toml input array

### PR DESCRIPTION
**Pull request by rm-dr**
This PR implements the feature detailed in #1100. Docs have been updated.

A final question I haven't had time to check:
Do all these fancy input file manipulations break latex error line numbers?

If they do, that's bad.

### CODE REVIEW

**Comment by pkgw** [comment]
I think the general direction here is good, but I think it would be very undesirable to break compatibility with existing `Tectonic.toml` documents so dramatically. Existing TOML fields should be honored with their current semantics. It's OK to have new fields that are incompatible with ones that already exist, but you should be able to put old files into a new program and get the same results, error-free.

This is a bit more subjective, but I think that the "main" file for a project should stay with the name `index.tex`, rather than changing it to `main.tex` (even though I just described it as the "main file" ...). Besides the backwards-compatibility aspect, I like the consistency with `index.html` and `index.js` seen in other environments.

**Comment by pkgw** [comment]
@rm-dr I noticed the same thing about the Harfbuzz stuff. It might help to rebase onto the latest `master`? I think that this branch might be based on an older commit before the update to Harfbuzz 8.3.0 ... which should then go away in the merge, but maybe git/GitHub is weird about that?

**Comment by rm-dr** [comment]
There we go, the git history now looks pretty.
This is almost done. There's just one thing left to do:

I need to change the default input files from the array to a single `main.tex` (or `index.tex`), and update the default workspace creation method. This isn't as easy as it sounds, since the v2 tests are tightly coupled to the file naming scheme, and fail if the defaults are changed.

File: docs/src/ref/tectonic-toml.md [hunk 1]
```diff
@@ -24,101 +24,52 @@ pubish = false
...
+#
+# Finally, you may include an "inline" document as follows:
+# inputs = [
+#   { inline = "\\documentclass[a4paper]{article}" },
+#   "main.tex"
+# ]
+# This will insert "\documentclass[a4paper]{article}" before main.tex
+# (with a newline), allowing you to set options without making a new file.
+inputs = "main.tex"
+```
```
Comment (pkgw) [inline docs/src/ref/tectonic-toml.md:75]: This new documentation structure is an improvement, I think. But the reference documentation should document the `preamble`, etc., fields, even if they're deprecated.

**Review by pkgw — COMMENTED** [review]
I have a few suggestions relating to the documentation, but I like how this is looking! I think this change will make a lot of people happy.

**Review by pkgw — APPROVED** [review]
D'oh, looks like this needs to be updated to not conflict with the curly-quotes change, but once that's done, I think you should merge this.

**Pull request merged**